### PR TITLE
Support Piping

### LINKED ISSUE #4: Support normal piping operation (`|`)

### PR DESCRIPTION
**Pull request by siliconlad**
image

### CODE REVIEW

**Comment by siliconlad** [comment]
Still a work in progress because I need to figure out how to run the builtin commands.

@CraigNewlands if you have any ideas I'm all ears.

**Comment by siliconlad** [comment]
Well this seems to be working:

image

File: src/command.rs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    if tokens.is_empty() {
+        Err("Tokens cannot be empty".into())
+    } else if is_builtin(&tokens[0]) {
+        Ok(Box::new(BuiltinCommand::new(tokens)?))
+    } else {
+        Ok(Box::new(ExternalCommand::new(tokens)?))
+    }
+}
+
+#[derive(Debug)]
```
Comment (CraigNewlands) [inline src/command.rs:29]: Out of interest what does this do?

File: src/traits.rs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+use std::error::Error;
+use std::fmt::Display;
+use std::process::ChildStdout;
+
+pub trait Runnable {
```
Comment (CraigNewlands) [inline src/traits.rs:5]: Are traits just templates?

**Review by CraigNewlands — APPROVED** [review]

File: src/traits.rs (hunk 2 above)
Comment (siliconlad) [reply src/traits.rs:5]: I think they're more equivalent to Interfaces (e.g. in Java). The book has a section on it [here](https://doc.rust-lang.org/book/ch10-02-traits.html).

File: src/command.rs (hunk 1 above)
Comment (siliconlad) [reply src/command.rs:29]: Let's you print the struct in a debug statement I think.

```rust
dbg!("{:?}")
```

**Pull request merged**